File: utils/browser_config.py
import os
import logging
from utils.sess_handler import SessionHandler


class Config:
    def __init__(self):
        self.browser = None
        self.page = None
        self.session_handler = None

    def is_headless(self):
        return os.getenv("headless") == "True"

    async def setup_browser(self, playwright):
        browser_type = os.getenv("BROWSER", "chromium")
        mode = os.getenv("mode")
        headless = self.is_headless()
        launch_args = {
            "headless": headless,
            "args": ["--start-maximized"]}

        if mode == 'pipeline':
            self.browser = await playwright[browser_type].launch(**launch_args)
        elif mode == 'local':
            self.browser = await playwright[browser_type].launch(**launch_args)
        elif mode == 'grid':
            server_url = "http://remote-playwright-server:4444"
            self.browser = await playwright[browser_type].connect(server_url)
        else:
            raise ValueError(f"Unsupported execution type: {mode}")

        self.session_handler = SessionHandler(self.browser, headless)

    async def context_init(self, storage_state=None, user_type="user"):
        context_options = {
            "viewport": {"width": 1920, "height": 1080} if self.is_headless() else None,
            "no_viewport": not self.is_headless()}

        if storage_state:
            context_options["storage_state"] = await self.session_handler.create_session(user_type)

        return await self.browser.new_context(**context_options)

    async def setup_page(self):
        context = await self.context_init()
        self.page = await context.new_page()
        return self.page

    async def setup_auth_page(self, auth_mode: str):
        context = await self.context_init(storage_state=True, user_type=auth_mode)
        self.page = await context.new_page()
        return self.page

    async def capture_handler(self):
        screenshot_option = os.getenv("screenshot", "off")
        if screenshot_option != "off" and self.page:
            screenshot_path = f"reports/screenshots/{await self.page.title()}.png"
            os.makedirs(os.path.dirname(screenshot_path), exist_ok=True)
            await self.page.screenshot(path=screenshot_path, full_page=True)


logging.getLogger('asyncio').setLevel(logging.WARNING)
logging.getLogger('filelock').setLevel(logging.CRITICAL)

File: utils/browserConfig.js
const { chromium, firefox, webkit } = require('playwright');
const fs = require('fs');
const path = require('path');
const SessionHandler = require('./sessHandler');
const LoginHandler = require('./loginHandler');

class BrowserConfig {
    constructor() {
        this.browser = null;
        this.page = null;
    }

    isHeadless() {
        return process.env.HEADLESS === 'true';
    }

    async setupBrowser() {
        const browserType = process.env.BROWSER || 'chromium';
const headless = this.isHeadless();
const launchOptions = { headless, args: ['--start-maximized'] };

const mode = process.env.MODE || 'local';  // Default to 'local' if MODE is not set

switch (mode) {
    case 'pipeline':
    case 'local':
        this.browser = await { chromium, firefox, webkit }[browserType].launch(launchOptions);
        break;
    case 'grid':
        const remoteUrl = 'http://remote-playwright-server:4444';
        this.browser = await { chromium, firefox, webkit }[browserType].connect(remoteUrl);
        break;
    default:
        throw new Error(`Unsupported execution type: ${mode}`);
}

    }

    async createPage(storageState = null, userType = 'user') {
        const contextOptions = {
            viewport: { width: 1920, height: 1080 },
        };
        if (storageState) {
            contextOptions.storageState = await this.createSession(userType);
        }
        const context = await this.browser.newContext(contextOptions);
        this.page = await context.newPage();
        return this.page;
    }

    async createSession(userType) {
        if (this.sessionHandler.isSessionExpired()) {
            const context = await this.browser.newContext();
            const page = await context.newPage();
            const login = new LoginHandler(page);
            const { username, password } = this.sessionHandler.getCredentials(userType);

            await login.login(username, password);
            const storageState = await context.storageState();
            this.sessionHandler.saveSession(storageState);
            await context.close();
        }
        return this.sessionHandler.sessionPath;
    }

    isSessionExpired(sessionPath) {
        const session = JSON.parse(fs.readFileSync(sessionPath));
        const currentTime = Date.now() / 1000;

        return session.cookies.some(cookie => cookie.expires <= currentTime);
    }

    getCredentials(userType) {
        const credentialsPath = path.resolve('data/credentials.json');
        if (!fs.existsSync(credentialsPath)) {
            throw new Error(`Credentials file not found: ${credentialsPath}`);
        }

        const credentials = JSON.parse(fs.readFileSync(credentialsPath));
        if (!credentials[userType]) {
            throw new Error(`User type '${userType}' not found in credentials`);
        }

        return credentials[userType];
    }
}

module.exports = BrowserConfig;
